libs/platform/ebpf_object.c

```diff
@@ -22,59 +22,137 @@ static ebpf_lock_t _ebpf_object_tracking_list_lock = {0};
  * 2) A pinning table entry holds a reference on it.
  * 3) A link object can hold a reference when it is associated with a hook.
  *
- * A link object can have a reference due to on of the following:
+ * A link object can have a reference due to one of the following:
  * 1) An open handle holds a reference on it.
  * 2) A pinning table entry holds a reference on it.
  *
+ * Libbpf has APIs like bpf_prog_get_next_id, bpf_map_get_next_id,
+ * and bpf_link_get_next_id to enumerate object IDs so every
+ * object gets a unique ID upon creation.
  */
-static _Requires_lock_held_(&_ebpf_object_tracking_list_lock) ebpf_list_entry_t _ebpf_object_tracking_list;
 
-static void
-_ebpf_object_tracking_list_insert(ebpf_object_t* object)
+typedef struct _ebpf_id_entry
+{
+    // Counter incremented each time a new object is stored here.
+    uint16_t counter;
+
+    // Pointer to object.
+    ebpf_object_t* object;
+} ebpf_id_entry_t;
+
+// Currently we allow a maximum of 1024 objects (links, maps,
+// and programs combined).  This can be increased in the future
+// if needed by making each array element store a linked list of
+// entries in order by ID, rather than a single entry, in which
+// case this becomes a hash table.
+static _Requires_lock_held_(&_ebpf_object_tracking_list_lock) ebpf_id_entry_t _ebpf_id_table[1024];
```

Comment:
or allocate a bigger array if there are more than 1024 objects?

Reply:
Maybe, but that wouldn't solve the problem, it would just punt it father down the road.  By the time 1024 becomes a problem, we probably won't want to punt it given there is an alternative.